include/cifer/abe/dippe.h

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2018 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CIFER_DIPPE_H
+#define CIFER_DIPPE_H
+
+#include <gmp.h>
+#include <amcl/fp12_BN254.h>
+#include <amcl/ecp_BN254.h>
+#include <amcl/ecp2_BN254.h>
+
+#include "cifer/data/vec.h"
+#include "cifer/internal/errors.h"
+
+/**
+* \file
+* \ingroup abe
+* \brief This is a decentralized inner-product predicate encryption
+* scheme by Yan Michalevsky and Marc Joye: "Decentralized Policy-Hiding
+* Attribute-Based Encryption with Receiver Privacy"
+*/
+
+
+/**
+ * cfe_dippe_cipher represents the ciphertext of the DIPPE scheme
+ */
+typedef struct cfe_dippe_cipher {
+    FP12_BN254   C_prime;
+    ECP_BN254*   C0;
+    ECP_BN254*   Ci;
+} cfe_dippe_cipher;
+
+/**
+ * Clears the ciphertext and frees allocated memory
+ *
+ * @param cipher A pointer to a cfe_dippe_cipher struct
+ */
+void cfe_dippe_cipher_free(cfe_dippe_cipher *cipher);
+
+/**
+ * cfe_dippe_pub_key represents the public key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_pub_key {
+    ECP2_BN254   g2_sigma;
+	ECP_BN254*   g1_W_A;
+	FP12_BN254*  gt_alpha_A;
+} cfe_dippe_pub_key;
+
+/**
+ * Clears the public key and frees allocated memory
+ *
+ * @param pk A pointer to a cfe_dippe_pub_key struct
+ */
+void cfe_dippe_pub_key_free(cfe_dippe_pub_key *pk);
+
+/**
+ * cfe_dippe_sec_key represents the secret key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_sec_key {
+	mpz_t        sigma;
+	cfe_vec      alpha;
+	cfe_vec      W;
+} cfe_dippe_sec_key;
+
+/**
+ * Clears the secret key and frees allocated memory
+ *
+ * @param sk A pointer to a cfe_dippe_sec_key struct
+ */
+void cfe_dippe_sec_key_free(cfe_dippe_sec_key *sk);
+
+/**
+ * cfe_dippe_sec_key represents the user secret key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_user_secret_key {
+    ECP2_BN254* Ki;
+} cfe_dippe_user_secret_key;
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk);
+
+/**
+ * cfe_dippe represents the DIPPE scheme
+ */
+typedef struct cfe_dippe {
+    unsigned int assump_size;
```

Comment:
`size_t` might be easier to read, also mostly used in the rest of `CiFEr`

Reply:
I replaced all `unsigned int` with `size_t`.